### PR TITLE
Abstract away extension handling

### PR DESCRIPTION
**Pull request by twsouthwick**
This removes the need for extensions to be too aware of the fact that they are themselves extensions. This incorporates a builder pattern to register configuration that can be added to by other extensions and provides simpler access patterns.

### CODE REVIEW

**Review by mjrousos — COMMENTED** [review]
As we discussed offline, I like the direction of abstracting extension infrastructure from components like steps and providers that consume options from them, but I found some of the options patterns hard to follow. It may be worth exploring if we can change FileOption in particular to make usage simpler.

**Review by mjrousos — APPROVED** [review]
Looks good. `FileOption` still exists and it's not clear to me why/whether it's necessary, but everything else looks good.

File: src/components/Microsoft.DotNet.UpgradeAssistant.Extensions/ExtensionServiceCollection.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        public IExtensionOptionsBuilder<TOption> AddExtensionOption<TOption>(string sectionName)
+            where TOption : class, new()
+        {
+            // Register options first so our custom factory isn't overwritten
+            Services.AddOptions();
+
+            // The default options factory cannot create an instance of ICollection<>
+            Services.AddTransient<IOptionsFactory<ICollection<TOption>>, CollectionOptionsFactory<TOption>>();
+            Services.AddTransient<IOptionsFactory<ICollection<FileOption<TOption>>>, CollectionOptionsFactory<FileOption<TOption>>>();
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/components/Microsoft.DotNet.UpgradeAssistant.Extensions/ExtensionServiceCollection.cs:21]: In order to map the files, we need to be able to get access to the file provider. Since we don't want to require the use of ifileoptions unless the user wants to, the FileOptions registration is kept to enable the mapping behind the scenes (it's an internal implementation). I thought there was a comment as to why it was there - I'll double check to make sure that's still there

**Review by mjrousos — APPROVED** [review]
LGTM
